packages/core/src/node/kill-tree.ts

```diff
@@ -0,0 +1,129 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import * as cp from 'child_process';
+import * as readline from 'readline';
+
+/**
+ * When this script is invoked as a standalone, the first argument must be the pid to kill the tree from.
+ */
+if (require.main === module) {
+    // Shield this process from erroneously exiting on SIGTERM, as this script might try to kill itself:
+    process.on('SIGTERM', () => process.exit(0));
+    const argv = process.argv.slice(2);
+    killTree(Number.parseInt(argv[0], 10));
+}
+
+/**
+ * @param ppid Process to kill along with its children (recursive).
+ */
+export async function killTree(ppid: number): Promise<void> {
+    if (process.platform === 'win32') {
+        await taskkillTree(ppid);
+    } else {
+        if (ppid === getpgid(ppid)) {
+            process.kill(-ppid);
+        } else {
+            for (const pid of await pgrepRecursive(ppid)) {
+                // Prevent killing the current process:
+                if (pid !== process.pid) {
+                    // Don't stop if a process fails to be killed (keep on killing the others):
+                    try {
+                        process.kill(pid);
+                    } catch (error) {
+                        console.error(error);
+                    }
+                }
+            }
+        }
+    }
+}
+
+/**
+ * Windows only.
+ *
+ * @param ppid The parent process to kill along with its children.
+ */
+function taskkillTree(ppid: number): Promise<void> {
+    return new Promise((resolve, reject) => {
+        const taskkill = cp.spawn('taskkill', ['/f', '/t', '/pid', ppid.toString(10)], { stdio: 'ignore' });
+        taskkill.on('error', reject);
+        taskkill.on('exit', (code, signal) => {
+            if (code === 0) {
+                resolve();
+            } else {
+                reject(new Error(`process exited with ${code ?? signal}`));
+            }
+        });
+    });
+}
+
+/**
+ * UNIX only.
+ *
+ * @returns the process group id for a given process.
+ */
+function getpgid(pid: number): number {
+    if (process.platform === 'win32') {
+        throw new Error('not supported');
+    }
+    const result = cp.spawnSync('ps', ['--pid', pid.toString(10), '--format', 'pgid='], { encoding: 'utf8' });
+    if (result.error) {
+        throw result.error;
+    }
+    if (result.status !== 0) {
+        throw new Error(`ps exited with ${result.status ?? result.signal}`);
+    }
+    return Number.parseInt(result.stdout.trim(), 10);
+}
+
+/**
+ * UNIX only.
+ *
+ * Call pgrep recursively since pgrep only lists direct children for a given pid.
+ *
+ * @param ppid The process id to fetch children from.
+ * @param pids All found pids (ppid included).
+ * @returns All found pids (ppid included).
+ */
+async function pgrepRecursive(ppid: number, pids: Set<number> = new Set()): Promise<Set<number>> {
```

Comment:
An alternative might be something like `ps xao pid,pgid` which produces a double column display with just process ID's and group ID's e.g.:

```
30802 30480
30810 30480
30947 30480
31006 30480
31020 31020
31163 30480
31169 30480
31194 30480
31211 30480
31239 30480
31328 30480
```

Reply:
I don't think we need the PGID here. Also this code used to be all synchronous, hence why I did not use `ps-tree`. Now that is is async I guess I can use it?

Reply:
Isn't the goal here to kill all processes with a specific group ID? Detaching the backend main process allows it to be the group leader (i.e. sets the group ID to its PID for all its descendants) and it is all and only the processes in that group that we want to kill with this code, isn't it? Or am I missing the mark?

Reply:
In this specific case `pgrepRecursive` was need in the case where we couldn't use the PGID. In such case, the idea was to manually fetch the process hierarchy following parent/child relationships, independently from the notion of groups.